### DISCUSSION TITLE
The calculated time series is 59 minutes and 59 seconds

### DISCUSSION THREAD

**Discussion by t298** [pyswmm/pyswmm#480] [Q&A] [answered]
I set the step length to be 1 hour, but I calculated it to be 59 minutes and 59 seconds
![image](https://github.com/pyswmm/pyswmm/assets/49806840/fde5aa29-9fe0-4ff2-b757-e3c18ca53e87)
![image](https://github.com/pyswmm/pyswmm/assets/49806840/2b214c6c-db45-4732-b563-b2a053c9b631)
[floodregulation.zip](https://github.com/pyswmm/pyswmm/files/12573630/floodregulation.zip)

**Comment by bemcdonnell** [comment]
@t298 moving to a discussion.  Are you saying you used sim.step_advance() and the time step isn’t what you’re expecting?

**Reply by t298** [reply]
I expect that the step length should be calculated once an hour, and there is no problem in other time periods, but the time after calculation of this time period has become this.

**Reply by bemcdonnell** [reply]
@t298 im struggling to understand the issue you’re having. Your “report step” is 1 hour which is what you can find in your OUT file. Your solver time step is <=30 seconds. I see you’re opening the OUT file - the dates are sorted at FLOAT. In the screenshot you sent, it looks like the dates are being converted perfectly. What is the problem?

**Reply by t298** [reply]
I'm terribly sorry, I've got the wrong picture
![image](https://github.com/pyswmm/pyswmm/assets/49806840/8b782dbd-8071-4be0-a654-ffaf70d29876)

**Comment by karosc** [comment]
@t298, I ran your model using swmm-toolkit 0.14.0 and pyswmm 1.4.0 and got the appropriate hourly timestamp values. Can you past the output of `pip list`?

**Reply by t298** [reply]
![image](https://github.com/pyswmm/pyswmm/assets/49806840/5d40e00c-e57d-4915-9dd1-9d14eda0ba62)
pyswmm 1.4.0，but swmm-toolkit 0.14.2

**Reply by karosc** [reply]
I'm hoping you can paste the output of your attached script in the terminal rather than pycharm screen clips. Is that possible?

**Reply by t298** [reply]
![image](https://github.com/pyswmm/pyswmm/assets/49806840/96f6713d-60dc-4bf5-9eeb-df06f8712623)

**Reply by t298** [reply]
![image](https://github.com/pyswmm/pyswmm/assets/49806840/63718406-6760-4f51-b4fb-46cb1d90622f)

**Reply by t298** [reply]
![image](https://github.com/pyswmm/pyswmm/assets/49806840/69457e53-c9c7-4604-986d-8d0f6a3ec14b)

**Reply by bemcdonnell** [reply]
@t298, does this happen with any simulation start date? For example, If you start it 1 day and 5 hours later ?

**Reply by t298** [reply]
Not all time periods have this error, but a not special time will have this problem, the same code, the same file, after I change the time, there will be no problem

**Comment by karosc** [comment, marked as answer]
@t298 , this might be a bug with the [official SWMM output file api](https://github.com/pyswmm/Stormwater-Management-Model/blob/main/src/outfile/swmm_output.c) or just simply a floating point arithmetic precision issue. The following code shows the issue:

```python
from swmm.toolkit import output
from julian import from_jd
handle = output.init()
out = output.open(handle, "floodregulation.out")
start_date = output.get_start_date(handle)
print(from_jd(start_date + 2415018.5))    # from 
#2019-09-08 00:59:59.999986
```

Because pyswmm derives timestamps from the report start and the report step, all the timestamps have the same micro second precision.

We might want to implement something [like this code](https://stackoverflow.com/a/49309848) or track down way the the dates and times are not coming out of SWMM at proper second intervals.

**Reply by t298** [reply]
So what can I do to avoid this type of error?

**Comment by karosc** [comment]
@bemcdonnell , I've done some testing and I now think this is an issue with [how the start_date is computed in pyswmm](https://github.com/pyswmm/pyswmm/blob/22aed805ec435dce93d0ccc084041e380827e01d/pyswmm/output.py#L179). Adding that big float then putting it through from_jd might be creating some precision error. Replacing that line with following code works well for me and would remove julian from our dependencies.

```python

from swmm.toolkit import output
from julian import from_jd
import datetime
handle = output.init()
out = output.open(handle, "floodregulation.out")


start_datetime= output.get_start_date(handle)

remainder = start_datetime%1
days = start_datetime - remainder 
seconds = remainder*86400
sd = datetime.datetime(1899,12,30) + datetime.timedelta(days = days , seconds = seconds)
print(sd)
#2019-09-08 01:00:00
```

**Comment by bemcdonnell** [comment]
@t298 before we fix this bug I’d just encourage you to add one second to the time steps +datetime.timedelta(seconds=1). I think you could write some simple logic to check the if seconds in your time step is >=30 and add a second. It’s a bit hacky but that should do the trick.

**Reply by t298** [reply]
ok,thanks
